### DISCUSSION TITLE
Preload Images

### DISCUSSION THREAD

**Discussion by randyperrone** [jspsych/jsPsych#1016] [Q&A] [answered]
I am loading 128 different images into an array as such 
```
images = ['img/h_c_m_hi_1.png', 'img/h_c_m_hi_2.png', img/h_c_m_hi_3.png', 'img/h_c_m_hi_4.png', etc...].  
```
I have a fixation, test trial and timeline as such: 

```
let fixation = {
    type: 'html-keyboard-response',
    stimulus: '<div style="font-size:' + FIXATION_SIZE + 'px;">' + FIXATION_KEY + '</div>',
    choices: jsPsych.NO_KEYS,
    trial_duration: function(){
        return jsPsych.randomization.sampleWithoutReplacement(FIXATION_DURATION, 1)[0];
    },
    data: { test_part: 'fixation' }
};
let test = {
    type: "image-keyboard-response",
    stimulus: ?,
    choices: [KEYBOARD_PRESS_HAPPY, KEYBOARD_PRESS_FEAR],
    stimulus_duration: STIMULUS_DURATION,
    trial_duration: TRIAL_DURATION,
    maintain_aspect_ration: MAINTAIN_IMG_ASPECT_RATIO,
    stimulus_height: STIMULUS_HEIGHT,
    stimulus_width: STIMULUS_WIDTH,
    post_trial_gap: POST_TRIAL_GAP
};
let test_procedure = {
    timeline: [fixation, test],
    preload_images: images,
    timeline_variables: images,
    repetitions: NUMBER_OF_TRIALS
};
timeline.push(test_procedure);
jsPsych.init({
    timeline: timeline
});
```

My question is what am I supposed to use for stimulus in test? Is there a way to set stimulus to just use any image, then set timeline_variables in test_procedure to a sequence of images that I want to use in a specific order (such as an ordered array of images)? I want to be able to randomize the array of 128 images in a specific way prior and then use that as the sequence of images in the test. I seem to be having trouble with the image preloader.  
Thanks so much for you help! -Randy

**Comment by becky-gilbert** [comment, marked as answer]
Hi @randyperrone! The [timeline_variables](https://www.jspsych.org/overview/timeline/#timeline-variables) array that you use in the test_procedure should be an array of objects, where each object represents a trial and contains the parameter-value pairs that are specific to that trial. So in your case, the timeline_variables array might have this structure:
```javascript
var trial_info = [
  {stim: 'img/h_c_m_hi_1.png'}, 
  {stim: 'img/h_c_m_hi_2.png'}
  etc.
];
``` 
Then, in your trial object, wherever you have a parameter value that should come from your timeline_variables array (instead of taking a fixed value), you can use the function `jsPsych.timelineVariable()` with the name of the corresponding key in your timeline_variables objects. I.e. `stimulus: jsPsych.timelineVariable('stim')`.

Also, the image preloading option is actually part of the [`jsPsych.init` settings](https://www.jspsych.org/core_library/jspsych-core/#jspsychinit), rather that part of a trial procedure, so you'll just need to move `preload_images: images` into `jsPsych.init`. 

So here's how it looks all together:
```javascript
var trial_info = [
  {stim: 'img/h_c_m_hi_1.png'}, 
  {stim: 'img/h_c_m_hi_2.png'}
];
let fixation = {
    type: 'html-keyboard-response',
    stimulus: '<div style="font-size:' + FIXATION_SIZE + 'px;">' + FIXATION_KEY + '</div>',
    choices: jsPsych.NO_KEYS,
    trial_duration: function(){
        return jsPsych.randomization.sampleWithoutReplacement(FIXATION_DURATION, 1)[0];
    },
    data: { test_part: 'fixation' }
};
let test = {
    type: "image-keyboard-response",
    stimulus: jsPsych.timelineVariable('stim'), // this now comes from each trial object in your trial_info array
    choices: [KEYBOARD_PRESS_HAPPY, KEYBOARD_PRESS_FEAR],
    stimulus_duration: STIMULUS_DURATION,
    trial_duration: TRIAL_DURATION,
    maintain_aspect_ration: MAINTAIN_IMG_ASPECT_RATIO,
    stimulus_height: STIMULUS_HEIGHT,
    stimulus_width: STIMULUS_WIDTH,
    post_trial_gap: POST_TRIAL_GAP
};
let test_procedure = {
    timeline: [fixation, test],
    timeline_variables: trial_info, // use the trial_info array here instead of the array of images
    repetitions: NUMBER_OF_TRIALS
};
timeline.push(test_procedure);
jsPsych.init({
    timeline: timeline,
    preload_images: images // image preloading goes here
});
```
I hope that helps. Please let us know if that's not what you wanted or have any other questions.

**Reply by randyperrone** [reply]
Thanks for you reply! So I am trying to move away from hard coding 128 objects and am hoping there is some easier way.  I originally had
```
let h_i_m_hi_1 = {stimulus: "img/h_i_m_hi_1.png",data: { test_part: 'test', correct_response: KEYBOARD_PRESS_HAPPY }}
let h_i_m_hi_2 = {stimulus: "img/h_i_m_hi_2.png",data: { test_part: 'test', correct_response: KEYBOARD_PRESS_HAPPY }}
let h_i_m_hi_3 = {stimulus: "img/h_i_m_hi_3.png",data: { test_part: 'test', correct_response: KEYBOARD_PRESS_HAPPY }}
let h_i_m_hi_4 = {stimulus: "img/h_i_m_hi_4.png",data: { test_part: 'test', correct_response: KEYBOARD_PRESS_HAPPY }}
etc... all the way to 128 of these!
```
and set the sequence to 
```
sequence = [h_i_m_ca_4, f_i_m_as_2, h_c_m_as_4, f_i_m_as_3, h_c_f_af_4, f_i_m_ca_1, f_c_m_af_4, h_c_m_af_2, f_c_f_ca_2, f_i_f_hi_1, h_i_f_as_2, f_i_f_af_4, f_c_f_af_3, h_c_f_hi_1, h_i_f_ca_1, h_c_m_hi_1, f_i_f_hi_4, h_i_f_af_4, f_i_m_hi_2, h_c_m_af_4, f_c_f_af_1, h_c_m_ca_3, f_c_f_as_1, h_i_f_as_3, f_i_m_af_4, f_c_f_hi_3, h_i_m_hi_2, f_i_f_hi_3, h_c_m_ca_4, h_i_m_af_2, h_c_f_as_4, h_i_m_af_1, h_c_f_af_2, h_i_f_as_4, f_i_f_ca_3, h_i_f_hi_3, f_c_f_ca_1, f_i_m_hi_4, f_c_f_af_2, f_i_f_as_3, f_c_m_as_3, h_i_m_af_3, f_c_m_as_4, f_i_m_hi_3, h_c_f_as_1, h_i_m_as_2, f_c_m_ca_3, f_i_f_as_2, f_c_f_hi_2, h_i_f_as_1, f_c_m_af_2, f_i_f_af_3, h_i_f_ca_4, h_c_f_hi_2, f_i_f_ca_2, h_c_f_af_3, f_i_f_ca_1, f_c_m_as_1, h_i_f_ca_2, h_c_f_ca_1, f_c_m_ca_4, h_c_m_as_2, f_i_f_ca_4, h_i_m_af_4, h_c_m_ca_2, h_i_f_hi_4, f_c_f_hi_1, h_i_f_af_1, f_c_m_hi_3, f_i_m_ca_2, h_i_m_as_4, f_i_m_ca_4, h_c_m_hi_4, f_i_f_as_4, h_i_f_hi_2, h_c_m_as_3, f_c_m_ca_1, h_c_f_hi_4, f_c_f_ca_3, h_i_m_ca_2, h_c_f_ca_2, h_i_m_as_1, f_c_m_hi_2, h_c_m_as_1, f_c_f_hi_4, f_i_m_as_4, h_c_f_hi_3, h_i_f_ca_3, f_c_f_ca_4, f_i_m_as_1, h_i_m_ca_1, f_i_f_hi_2, h_c_f_as_3, f_c_m_af_3, h_i_f_af_3, f_c_f_as_3, h_i_m_as_3, f_i_m_af_2, h_c_f_as_2, h_i_m_hi_4, h_c_f_ca_3, h_i_m_hi_1, h_c_m_af_1, f_c_m_as_2, h_i_m_hi_3, f_i_m_ca_3, h_c_m_hi_3, f_c_f_af_4, h_c_f_ca_4, h_i_m_ca_3, h_c_m_hi_2, h_i_f_hi_1, f_c_m_ca_2, f_i_f_af_1, h_c_m_ca_1, f_c_m_hi_1, h_i_f_af_2, h_c_m_af_3, f_i_f_as_1, f_c_m_hi_4, f_i_m_af_3, h_c_f_af_1, f_c_m_af_1, f_i_m_hi_1, f_c_f_as_2, f_i_m_af_1, f_c_f_as_4, f_i_f_af_2];
```
but that is extremely error prone.  I then set the stimulus to 
```
stimulus: jsPsych.timelineVariable('stimulus'),
```
and 
```
let test_procedure = {
    timeline: [fixation, test],
    timeline_variables: sequence,
    repetitions: NUMBER_OF_TRIALS
};
```
which all worked fine.  It was just all hardcoded in there. Is there not an easier way to pull in 128 images without creating an object for each one? I created the images array with a for loop and am building the image path for each image dynamically into that images array which gave me:
```
images = ['img/h_c_m_hi_1.png', 'img/h_c_m_hi_2.png', img/h_c_m_hi_3.png', 'img/h_c_m_hi_4.png', etc...];
```
I'm just wondering if there is a way to set the images array above to be my image order without creating an object for each one, or do I just have to create an object for each one like you suggested and how I was doing it before when I hardcoded each one? Hope that wasn't too confusing!
Thanks again!
-Randy

**Reply by becky-gilbert** [reply]
>  Is there not an easier way to pull in 128 images without creating an object for each one? 

> I'm just wondering if there is a way to set the images array above to be my image order without creating an object for each one

I don't know the answer to this, but someone else like @jodeleeuw or @vijaymarupudi might know. 

If you're already looping through the number of images to put them into an array for preloading, then you could also dynamically create your timeline_variables array in the same loop, e.g.
```javascript
images.push(this_img); // add to images array for preloading
trial_info.push({stim: this_img}); // create object and add to trial_info array for timeline_variables
```

**Reply by randyperrone** [reply]
Awesome! That works! Thanks for the help!
